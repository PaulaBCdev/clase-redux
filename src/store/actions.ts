import type { AppThunk } from ".";
import { login } from "../pages/auth/service";
import type { Credentials } from "../pages/auth/types";
import {
  createTweet,
  getLatestTweets,
  getTweet,
} from "../pages/tweets/service";
import type { Tweet } from "../pages/tweets/types";

type AuthLoginPending = {
  type: "auth/login/pending";
};

type AuthLoginFulfilled = {
  type: "auth/login/fulfilled";
};

type AuthLoginRejected = {
  type: "auth/login/rejected";
  payload: Error;
};

type AuthLogout = {
  type: "auth/logout";
};

type TweetsLoadedFulfilled = {
  type: "tweets/loaded/fulfilled";
  payload: Tweet[];
};

type TweetsCreatedFulfilled = {
  type: "tweets/created/fulfilled";
  payload: Tweet;
};

type UiResetError = {
  type: "ui/reset-error";
};

export const authLoginPending = (): AuthLoginPending => ({
  type: "auth/login/pending",
});

export const authLoginFulfilled = (): AuthLoginFulfilled => ({
  type: "auth/login/fulfilled",
});

export const authLoginRejected = (error: Error): AuthLoginRejected => ({
  type: "auth/login/rejected",
  payload: error,
});

// esta accion se va a encargar de juntar en un mismo sitio los authLogin fulfilled, pending y rejected
export function authLogin(credentials: Credentials): AppThunk<Promise<void>> {
  // esta funcion me va a devolver un thunk, que a su vez me va a devolver una promesa de void (nada)
  return async function (dispatch) {
    dispatch(authLoginPending());
    try {
      await login(credentials);
      dispatch(authLoginFulfilled());
    } catch (error) {
      if (error instanceof Error) {
        dispatch(authLoginRejected(error));
      }
      throw error;
    }
  };
}
// IMPORTANTE: fijate que esta accion devuelve una funcion. Las acciones lo que suelen devolver son objetos, ya que es lo que pide dispatch. Sin embargo, cuando se ejecute authLogin(), el thunk (middleware) va a ver que es una funcion y la va a ejecutar antes de pasarsela al dispatch (de hecho, fijate que hace los dispatch dentro de la funcion, y a cada dispatch le pasamos el objeto debido dependiendo del momento: empieza con un pending; manda las credenciales al servicio del login y si son correctas hace dispatch de fulfilled, y si son erroneas, hace dispatch de rejected)

export const authLogout = (): AuthLogout => ({
  type: "auth/logout",
});

export const tweetsLoadedFulfilled = (
  tweets: Tweet[],
): TweetsLoadedFulfilled => ({
  type: "tweets/loaded/fulfilled",
  payload: tweets,
});

export const tweetsCreatedFulfilled = (
  tweet: Tweet,
): TweetsCreatedFulfilled => ({
  type: "tweets/created/fulfilled",
  payload: tweet,
});

export function tweetsLoaded(): AppThunk<Promise<void>> {
  return async function (dispatch, getState) {
    const state = getState();
    if (state.tweets) {
      return;
    }
    // El crear una variable state que es el estado actual que hay guardado en Redux sirve para que, en el caso de que ya haya un array (es decir, ya se ha llamado a la api y se han obtenido los tweets) no se vuelva a realizar otra llamada a la api

    try {
      // Igual que hicimos con la funcion authLogin(), aqui podemos tratar un tweetsLoadedPending (los tweets estan cargando)...
      const tweets = await getLatestTweets();
      dispatch(tweetsLoadedFulfilled(tweets)); //Aqui cogemos los tweets que se han cargado de la API y se los mandamos al estado global para que Redux los guarde
    } catch (error) {
      // ... y aqui un tweetsLoadedRejected (error al cargar os tweets)
      console.log(error);
    }
  };
}

export function tweetsCreate(tweetContent: string): AppThunk<Promise<Tweet>> {
  return async function (dispatch) {
    try {
      // Aqui se trataria el tweetsCreatedPending
      // Aqui se trata el tweetsCreatedFulfilled
      const createdTweet = await createTweet(tweetContent);
      const tweet = await getTweet(createdTweet.id.toString()); // createdTweet no contiene toda la informacion que pide sparrest, por lo que daria un error. Creamos una variable tweet que ya si contiene toda la informacion de tweet que necesita sparrest. ATENCION: este paso no es necesario en la practica, nos vale solo con el createdTweet
      dispatch(tweetsCreatedFulfilled(tweet));
      return tweet;
    } catch (error) {
      // Aqui se trataria el tweetsCreatedRejected
      console.log(error);
      throw error;
    }
  };
}

export const uiResetError = (): UiResetError => ({
  type: "ui/reset-error",
});

export type Actions =
  | AuthLoginPending
  | AuthLoginFulfilled
  | AuthLoginRejected
  | AuthLogout
  | TweetsCreatedFulfilled
  | TweetsLoadedFulfilled
  | UiResetError;
