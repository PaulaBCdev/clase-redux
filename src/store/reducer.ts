import type { Tweet } from "../pages/tweets/types";
import type { Actions, ActionsRejected } from "./actions";

export type State = {
  auth: boolean;
  tweets: {
    loaded: boolean;
    data: Tweet[];
  };
  ui: {
    pending: boolean;
    error: Error | null;
  };
};

const defaultState: State = {
  auth: false,
  tweets: {
    loaded: false,
    data: [],
  },
  ui: {
    pending: false,
    error: null,
  },
};

/* export function reducer(state = defaultState, action: Actions): State {
  switch (action.type) {
    case "auth/login":
      return { ...state, auth: true };
    case "auth/logout":
      return { ...state, auth: false };
    case "tweets/created":
      return { ...state, tweets: [...state.tweets, action.payload] };
    case "tweets/loaded":
      return { ...state, tweets: action.payload };
    default:
      return state;
  }
} */

export function auth(
  state = defaultState.auth,
  action: Actions,
): State["auth"] {
  switch (action.type) {
    case "auth/login/fulfilled":
      return true;
    case "auth/logout":
      return false;
    default:
      return state;
  }
}

export function tweets(
  state = defaultState.tweets,
  action: Actions,
): State["tweets"] {
  switch (action.type) {
    case "tweets/loaded/fulfilled":
      return { loaded: true, data: action.payload };
    case "tweets/detail/fulfilled":
      return { loaded: false, data: [action.payload] };
    case "tweets/created/fulfilled":
      return { ...state, data: [action.payload, ...(state.data ?? [])] };
    // (state ?? []) quiere decir que si llegados a este punto state es null, me carga un array vacio
    default:
      return state;
  }
}

function isRejected(action: Actions): action is ActionsRejected {
  return action.type.endsWith("/rejected");
}

export function ui(state = defaultState.ui, action: Actions): State["ui"] {
  if (action.type === "auth/login/pending") {
    return { pending: true, error: null };
  }

  if (action.type === "auth/login/fulfilled") {
    return { pending: false, error: null };
  }

  if (isRejected(action)) {
    return { pending: false, error: action.payload };
  }

  if (action.type === "ui/reset-error") {
    return { ...state, error: null };
  }

  return state;
}

/* export function reducer(state = defaultState, action: Actions): State {
  return {
    auth: auth(state.auth, action),
    tweets: tweets(state.tweets, action),
  };
} */

/* ESTO MEJOR SE HACE EN EL INDEX.TS
export const reducer = combineReducers({
  auth, // igual que auth: auth (en auth me ejecutas la funcion auth)
  tweets, // igual que tweets: tweets (en tweets me ejecutas la funcion tweets)
});
A combineReducers se le pasa un objeto con la misma estructura que la indicada en el tipo, en este caso, State
 */
