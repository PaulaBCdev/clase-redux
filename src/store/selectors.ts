// Aqui van todas las funciones que voy a necesitar para sacar la informacion que necesite de Redux

import type { RootState } from ".";

export const getIsLogged = (state: RootState) => state.auth;

export const getTweets = (state: RootState) => state.tweets.data;

/* export const getTweetDetail = (state: RootState, tweetId?: string) =>
  state.tweets.find((tweet) => tweet.id === Number(tweetId)); */
export function getTweetDetail(tweetId?: string) {
  //A partir de aqui seria el selector real, la linea de antes crea una funcion que llama al selector
  return function (state: RootState) {
    return state.tweets.data.find((tweet) => tweet.id === Number(tweetId));
  };
}

export const getUi = (state: RootState) => state.ui;
