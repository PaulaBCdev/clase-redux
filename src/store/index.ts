import { combineReducers, createStore } from "redux";
import * as reducers from "./reducer.ts";
import { devToolsEnhancer } from "@redux-devtools/extension";
import { useDispatch, useSelector } from "react-redux";

const rootReducer = combineReducers(reducers);

export default function configureStore(
  preloadedState: Partial<reducers.State>,
) {
  const store = createStore(
    rootReducer,
    preloadedState as never, // me dice si de base estoy logeada o no
    /* // @ts-expect-error: import devtools extension
    window.REDUX_DEVTOOLS_EXTENSION &&
      // @ts-expect-error: import devtools extension
      window.REDUX_DEVTOOLS_EXTENSION(), */
    devToolsEnhancer(),
  );
  return store;
}

export type AppStore = ReturnType<typeof configureStore>; // Tipo del store (lo que retorna configureStore)
export type AppGetState = AppStore["getState"]; // Tipo de la funcion .getState() del store
export type RootState = ReturnType<AppGetState>; // Tipo del estado completo del store
export type AppDispatch = AppStore["dispatch"]; // Tipo de la funcion dispatch() del store

export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();
