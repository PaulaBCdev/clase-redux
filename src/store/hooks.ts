import { useAppDispatch, useAppSelector } from ".";
import { authLogin, authLogout } from "./actions";
import { getIsLogged } from "./selectors";

export function useAuth() {
  return useAppSelector(getIsLogged);
}

export function useLogin() {
  const dispatch = useAppDispatch();
  return function () {
    return dispatch(authLogin());
  };
}
// Esta funcion, primero crea una variable dispatch para no tener que escribir el pedazo de nombre de useAppDispatch; se va a usar paraque se ejecute la accion que le indiquemos.
// La funcion devuelve otra funcion que hace que dispatch llame a la accion authLogin(), que esta en el archivo actions.ts, donde se ve que devuelve un type: "auth/login". Como ese es el type que recoge el case del switch de la funcion auth() (en el fichero reducer.ts), provoca que auth() devuelva un true; por lo tanto, la clave auth de rootReducer del fichero index.js (la variable que se crea con el combineReducers) ahora tiene el valor true, lo que indica que el usuario esta logeado.

export function useLogout() {
  const dispatch = useAppDispatch();
  return function () {
    return dispatch(authLogout());
  };
}
