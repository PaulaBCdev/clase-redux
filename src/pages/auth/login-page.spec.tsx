import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import LoginPage from "./login-page";
import { Provider } from "react-redux";
import type { RootState } from "../../store";
import { authLogin, uiResetError } from "../../store/actions";

//MOCKEAR UN MODULO
vi.mock("../../store/actions");
// asi podemos testear si se llama al modulo sin tener que hacer la llamada de verdad

describe("LoginPage", () => {
  const state: RootState = {
    auth: false,
    tweets: { data: [], loaded: false },
    ui: {
      pending: false,
      error: null,
    },
  };
  const renderComponent = (error?: Error) => {
    if (error) {
      state.ui.error = error;
    }

    return render(
      <Provider
        store={{
          getState: () => state,
          // @ts-expect-error: subscribe
          subscribe: () => {},
          // @ts-expect-error: dispatch
          dispatch: () => {},
        }}
      >
        <LoginPage />
      </Provider>,
    );
  };

  test("should render", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test("should dispatch login action", async () => {
    renderComponent();

    const usernameInput = screen.getByLabelText(/username/);
    const passwordInput = screen.getByLabelText(/password/);

    const button = screen.getByRole("button");

    expect(button).toHaveTextContent("Log in");
    expect(button).toBeDisabled();

    /* fireEvent.change(usernameInput, { target: { value: "paula" } });
    fireEvent.change(passwordInput, { target: { value: "1234" } }); */
    await userEvent.type(usernameInput, "paula");
    await userEvent.type(passwordInput, "1234");

    expect(button).toBeEnabled();

    /* fireEvent.click(button); */
    await userEvent.click(button);

    expect(authLogin).toHaveBeenCalledWith({
      username: "paula",
      password: "1234",
    });
  });

  test("should render error", async () => {
    const error = new Error("Wrong username/password");
    const { container } = renderComponent(error);

    expect(container).toMatchSnapshot();

    const alert = screen.getByRole("alert");
    expect(alert).toHaveTextContent(error.message);

    await userEvent.click(alert);
    expect(uiResetError).toHaveBeenCalled();
  });
});
