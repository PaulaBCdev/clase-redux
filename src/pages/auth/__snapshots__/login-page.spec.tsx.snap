// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage > should render 1`] = `
<div>
  <div
    class="login-page"
  >
    <h1
      class="login-page-title"
    >
      Log in to Twitter
    </h1>
    <form>
      <div
        class="form-field"
      >
        <label
          class="form-field-label"
        >
          <span>
            phone, email or username
          </span>
          <input
            autocomplete="off"
            class="form-field-input"
            name="username"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="form-field"
      >
        <label
          class="form-field-label"
        >
          <span>
            password
          </span>
          <input
            autocomplete="off"
            class="form-field-input"
            name="password"
            type="password"
            value=""
          />
        </label>
      </div>
      <button
        class="sc-bXdtCk DqBej login-form-submit"
        disabled=""
        type="submit"
      >
        Log in
      </button>
    </form>
  </div>
</div>
`;

exports[`LoginPage > should render error 1`] = `
<div>
  <div
    class="login-page"
  >
    <h1
      class="login-page-title"
    >
      Log in to Twitter
    </h1>
    <form>
      <div
        class="form-field"
      >
        <label
          class="form-field-label"
        >
          <span>
            phone, email or username
          </span>
          <input
            autocomplete="off"
            class="form-field-input"
            name="username"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="form-field"
      >
        <label
          class="form-field-label"
        >
          <span>
            password
          </span>
          <input
            autocomplete="off"
            class="form-field-input"
            name="password"
            type="password"
            value=""
          />
        </label>
      </div>
      <button
        class="sc-bXdtCk DqBej login-form-submit"
        disabled=""
        type="submit"
      >
        Log in
      </button>
    </form>
    <div
      class="login-page-error"
      role="alert"
    >
      Wrong username/password
    </div>
  </div>
</div>
`;
